File: 剑指offer/7.重建二叉树.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        pos_map = {}
        for i, val in enumerate(inorder):
            pos_map[val] = i

        def help(preorder_left, preorder_right, inorder_left, inorder_right):
            if preorder_left > preorder_right:
                return None

            root_val = preorder[preorder_left]
            root = TreeNode(root_val)

            index = pos_map[root_val]

            root.left = help(preorder_left + 1,
                             index - inorder_left + preorder_left,
                             inorder_left, index - 1)
            root.right = help(index - inorder_left + preorder_left + 1,
                              preorder_right, index + 1, inorder_right)

            return root

        n = len(inorder)
        return help(0, n - 1, 0, n - 1)

File: 剑指offer/7.重建二叉树.cpp
#include <unordered_map>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x)
        : val(x)
        , left(NULL)
        , right(NULL)
    {
    }
};

class Solution {
public:
    unordered_map<int, int> mem;

    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int preorder_left, int preorder_right, int inorder_left, int inorder_right)
    {
        if (preorder_left > preorder_right) {
            return NULL;
        }

        int val = preorder[preorder_left];
        int inorder_root = mem[val];
        int left_len = inorder_root - inorder_left;

        TreeNode* root = new TreeNode(val);
        root->left = build(preorder, inorder, preorder_left + 1, preorder_left + left_len, inorder_left, inorder_root - 1);
        root->right = build(preorder, inorder, preorder_left + left_len + 1, preorder_right, inorder_root + 1, inorder_right);
        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder)
    {
        for (int i = 0; i < inorder.size(); i++) {
            mem[inorder[i]] = i;
        }

        return build(preorder, inorder, 0, preorder.size() - 1, 0, inorder.size() - 1);
    }
};